core/account-id/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+
+        prefix.ends_with('.') && &suffix[..suffix.len()] == parent_account_id.as_ref()
+    }
+
+    /// Returns true if the account ID length is 64 characters and it's a hex representation.
+    pub fn is_implicit(account_id: impl AsRef<str>) -> bool {
+        let account_id = account_id.as_ref();
+        account_id.len() == 64
+            && account_id.as_bytes().iter().all(|b| matches!(b, b'a'..=b'f' | b'0'..=b'9'))
+    }
+
+    /// Returns true if the account ID is suppose to be EVM machine.
+    pub fn is_evm(account_id: impl AsRef<str>) -> bool {
+        account_id.as_ref() == "evm"
+    }
+
+    /// Returns true if the account ID is the system account.
+    pub fn is_system(account_id: impl AsRef<str>) -> bool {
+        account_id.as_ref() == "system"
+    }
+
+    pub fn system_account() -> Self {
+        "system".parse().unwrap()
+    }
+
+    pub fn test_account() -> Self {
+        "test".parse().unwrap()
+    }
+}
+
+impl std::borrow::Borrow<str> for AccountId {
+    fn borrow(&self) -> &str {
+        self.as_ref()
+    }
+}
+
+impl FromStr for AccountId {
+    type Err = ParseAccountError;
+
+    fn from_str(account_id: &str) -> Result<Self, Self::Err> {
+        Self::validate(account_id)?;
+        Ok(Self(account_id.into()))
+    }
+}
+
+impl TryFrom<String> for AccountId {
+    type Error = ParseAccountError;
+
+    fn try_from(account_id: String) -> Result<Self, Self::Error> {
+        Self::validate(&account_id)?;
+        Ok(Self(account_id.into()))
+    }
+}
+
+impl From<AccountId> for String {
+    fn from(account_id: AccountId) -> Self {
+        account_id.0.into_string()
+    }
+}
+
+impl TryFrom<&[u8]> for AccountId {
+    type Error = ParseAccountError;
+
+    fn try_from(account_id: &[u8]) -> Result<Self, Self::Error> {
+        std::str::from_utf8(account_id).map_err(|_| Self::Error::Invalid)?.parse()
+    }
+}
+
+#[cfg(feature = "serde")]
+fn serde_validate_account_id<'de, D>(d: D) -> Result<Box<str>, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    let account_id = de::Deserialize::deserialize(d)?;
+    AccountId::validate(&account_id).map_err(de::Error::custom)?;
+    Ok(account_id)
+}
+
+#[cfg(feature = "borsh")]
+impl BorshDeserialize for AccountId {
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, std::io::Error> {
+        let account_id = BorshDeserialize::deserialize(buf)?;
+        Self::validate(&account_id)
+            .map_err(|err| std::io::Error::new(std::io::ErrorKind::InvalidData, err))?;
+        Ok(Self(account_id))
+    }
+}
+
+#[cfg(feature = "paperclip")]
+mod paperclip {
+    use super::AccountId;
+    use paperclip::v2::{models::DataType, schema::TypedData};
+    impl TypedData for AccountId {
+        fn data_type() -> DataType {
+            DataType::String
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    #[cfg(feature = "borsh")]
+    use borsh::{BorshDeserialize, BorshSerialize};
+    #[cfg(feature = "serde")]
+    use serde_json::json;
```

Comment:
unreleated to the PR, but this really wants fuzz tests, a-la the ones in semver: https://github.com/dtolnay/semver/tree/master/fuzz